Issue #240 - added support for "reject duplicate keys" config option

JohnTimm opened the pull request:
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

lukasj commented on impl/src/main/java/org/glassfish/json/JsonParserFactoryImpl.java:
```diff
@@ -41,17 +41,17 @@
     @Override
     public JsonParser createParser(InputStream in, Charset charset) {
-        return new JsonParserImpl(in, charset, bufferPool);
+        return new JsonParserImpl(in, charset, bufferPool, true);
```
Why not just adding new constructor and redirecting existing one to it, so the rest of the code does not have to be updated ?

JohnTimm replied:
I made the changes that you have recommended. Please see my latest commit.

lukasj commented on impl/src/main/java/org/glassfish/json/JsonParserImpl.java:
```diff
@@ -55,25 +56,29 @@
     private final JsonTokenizer tokenizer;
 
-    public JsonParserImpl(Reader reader, BufferPool bufferPool) {
+    public JsonParserImpl(Reader reader, BufferPool bufferPool, boolean allowDuplicateKeys) {
```
I'm not big fan of this approach - are we going to add new argument to the constructor for every property in the future? Having `Map<String,?> options` here instead would probably look better to me

JohnTimm replied:
I was following the existing code path / design. I have cleaned things up a bit. Please see my latest commit.

lukasj commented on impl/src/test/java/org/glassfish/json/tests/JsonReaderDuplicateKeyTest.java:
```diff
@@ -0,0 +1,40 @@
+package org.glassfish.json.tests;
```
missing license

JohnTimm replied:
Added license / copyright header to this file.

lukasj commented on impl/src/main/java/org/glassfish/json/JsonReaderImpl.java:
```diff
@@ -36,23 +36,28 @@
  * @author Jitendra Kotamraju
  */
 class JsonReaderImpl implements JsonReader {
+    public static final String ALLOW_DUPLICATE_KEYS = "org.glassfish.json.JsonReaderImpl.allowDuplicateKeys";
```
does it make sense to have `public static` field in a class with `package` level visibility?

JohnTimm replied:
I have added this configuration property to the JsonReader interface in the api module.

lukasj commented on impl/src/main/java/org/glassfish/json/JsonReaderImpl.java:
```diff
@@ -36,23 +36,28 @@
  * @author Jitendra Kotamraju
  */
 class JsonReaderImpl implements JsonReader {
+    public static final String ALLOW_DUPLICATE_KEYS = "org.glassfish.json.JsonReaderImpl.allowDuplicateKeys";
```
I'd change the property name to `org.glassfish.json.allowDuplicateKeys` instead and if the intention is to be available to users, consider defining this in the `api` package which is being exported in JPMS/OSGi

JohnTimm replied:
I made the change that you have recommended. Please see my latest commit.

lukasj commented on impl/src/main/java/org/glassfish/json/JsonParserImpl.java:
```diff
@@ -55,25 +56,29 @@
     private final Stack stack = new Stack();
     private final JsonTokenizer tokenizer;
 
-    public JsonParserImpl(Reader reader, BufferPool bufferPool) {
```
if this is reimplemented to sth like:
```
public JsonParserImpl(Reader reader, BufferPool bufferPool) {
    this(reader, bufferPool, true);
}
```
most changes done could have been avoided

JohnTimm replied:
I made the change that you have recommended. Please see my latest commit.

leadpony commented:
@JohnTimm  @lukasj 
* `JsonReaferFactory#getConfigInUse()` must return all configuration properties it recognized.
* I feel `FORBID_DUPLICATE_KEYS` is better naming than `ALLOW_DUPLICATE_KEYS`, because the property does not exist in the configuration properties by default and such default state means _allow duplicate keys_.

JohnTimm commented:
> @JohnTimm @lukasj
> 
>     * `JsonReaferFactory#getConfigInUse()` must return all configuration properties it recognized.
> 
>     * I feel `FORBID_DUPLICATE_KEYS` is better naming than `ALLOW_DUPLICATE_KEYS`, because the property does not exist in the configuration properties by default and such default state means _allow duplicate keys_.

I renamed `ALLOW_DUPLICATE_KEYS` to `FORBID_DUPLICATE_KEYS` and updated the logic accordingly. Please see my latest commit.

lukasj commented on api/src/main/java/jakarta/json/JsonReader.java:
```diff
@@ -52,6 +52,12 @@
+     * Configuration property to forbid duplicate keys. The value of the property could be
+     * be anything.
+     */
+    public static final String FORBID_DUPLICATE_KEYS = "jakarta.json.JsonReader.forbidDuplicateKeys";
```
Not here, this is jakarta API. I ment `org.glassfish.json.api` package - implementation specific API

JohnTimm replied:
ah, okay. I'll make that change.

JohnTimm replied:
@lukasj how would you like me to expose it from that package? In some sort of ConfigProperties interface?

JohnTimm replied:
I created a new interface named `JsonConfig` in the `org.glassfish.json.api` package and moved the `FORBID_DUPLICATE_KEYS` constant there. Please see my latest commit.

JohnTimm commented:
> Make sure that the copyright year is updated to 2020 in all files you changed.

Done. Build passed.

JohnTimm commented:
@lukasj @m0mus Is there a chance for this PR to make it into the next release candidate?

JohnTimm commented:
> Make sure that the copyright year is updated to 2020 in all files you changed.

Your requested changes have been addressed. What do we need to do to move this PR forward?

nmittles commented:
@m0mus Hi Dmitry. Is there anything holding this PR out? It looks like John has made all the requested changes.

JohnTimm commented:
@lukasj @m0mus This has been sitting here for months. Please let me know if more needs to be done before it can be merged.

lukasj commented:
We did not want this in 2.0.0 and currently waiting for master to open up for integrations (expect ~2-3 weeks)

JohnTimm commented:
@lukasj Now that Jakarta EE 8 has shipped, where do we stand on getting this PR pulled into master?

lmsurpre commented:
Bumped into this interesting article and it reminded me that we've been waiting on this pull request for 9 months:  https://labs.bishopfox.com/tech-blog/an-exploration-of-json-interoperability-vulnerabilities

lukasj reviewed (commented):
Shouldn't `JsonObjectBuilderImpl` be the one to decide if the to be added `key` should be rejected or not? Or is it expected that `parser.getObject()` may reject the field while `json.createObjectBuilder().add("firstName", "John").add("firstName", "Peter".build()` is OK to not throw any errors when the reject option is set to true? This does not consistent  to me.

JohnTimm commented:
@lukasj Everything is done through JsonObjectBuilder now so that the behavior should be consistent. Please take a look at my recent commits. We'd like to get this functionality merged sooner than later.

lukasj reviewed (approved):
thanks!

m0mus reviewed (approved):
LGTM
